Optimization v3 (#507)
mlr-org/mlr3book: book/optimization.qmd

This function allows us to construct a learner in the usual way but to leave the hyperparameters of interest to be unspecified within a set range.
This is best demonstrated by example:

```{r optimization-006}
learner = lrn("classif.svm",
  cost  = to_tune(1e-1, 1e5),
  gamma = to_tune(1e-1, 1),
  type  = "C-classification",
  kernel = "radial"
)
learner
```

Here we have constructed a classification SVM by setting the type to "C-classification", the kernel to "radial", and not fully specifying the `cost` and `gamma` hyperparameters but instead indicating that we will tune these parameters.

> The section ... and ... give

Author: Forgot the ref

::: {.callout-note}
The `cost` and `gamma` hyperparameters are usually tuned on the logarithmic scale.
You can find out more in @sec-logarithmic-transformations.
:::

Search spaces are usually chosen by experience.
In some cases these can be quite complex.

> The sections X and Y give a more detailed insight into the creation of tuning spaces.

Author: Added ref

instance$result
```

::: {.callout-note}
The measured performance is different from the previous section because not the same train-test splits were used.
The train-test splits are generated at the beginning of the optimization with the current seed (see @sec-resampling-inst).
:::

> 

Author: @larskotthoff @berndbischl Your suggestion

> 

Reviewer: I think there's value in at least briefly mentioning this.

This way it is possible to tune continuous parameters with discrete values, or to give transformations or dependencies.
One could, for example, tune the `cost` as above on three given special values, and introduce a dependency of `shrinking` on it.
Notice that a short form for `to_tune(<levels>)` is a short form of `to_tune(p_fct(<levels>))`.

:::{.callout-note}
When introducing the dependency, we need to use the `degree` value from *before* the implicit transformation, which is the name or `as.character()` of the respective value, here `"val2"`!
:::

```{r optimization-048}
learner$param_set$set_values(
  cost = to_tune(c(val1 = 0.3, val2 = 0.7)),
  shrinking = to_tune(p_lgl(depends = cost == "val2"))
)
learner$param_set$search_space()
```

The search space picks up dependencies from the underlying `r ref("ParamSet")` automatically.
So if the `kernel` is tuned, then `degree` automatically gets the dependency on it, without us having to specify that.

```{r optimization-049}
learner$param_set$values$kernel = to_tune(c("polynomial", "radial"))
learner$param_set$search_space()
```

> It is even possible to define whole `r ref("ParamSet")`s that get tuned over for a single parameter.

Reviewer 2: Ok this is not necessary because it is decently explained I think

Note here we use the term *unbiased* to refer only to the statistical procedure of the performance estimation.
The underlying prediction of the model could still be biased e.g. due to a bias in the data set.

```{r optimization-029}
rr$aggregate()
```

As a final note, nested resampling is computationally expensive, as a simple example using five outer folds and three inner folds with a grid search of resolution 5 used to tune 2 parameters, results in 5*3*5*5 = 375 iterations of model training/testing. In practice, you may often see closer to three folds used in inner resampling or even holdout, or if you have the resources then we recommend parallelization (@sec-parallelization).

## Advanced Tuning

This section is devoted to advanced tuning techniques.

> The topics are not necessary for a basic understanding of tuning but make tuning more efficient and robust.

Reviewer 3: @be-marc please add the optional section tag or don't and resolve this conversation instead

### Encapsulation and Fallback Learner {#sec-encapsulation-fallback}

So far, we have only looked at the case where no issues occur.

> However, it often happens that learners with certain configurations do not converge, run out of memory, or terminate with an error.

Reviewer 3: I'm merging this suggestion as it's literally repeating the line before.

Since the tuner automatically suggests configurations, we cannot completely control the process and there should be no interruptions.
We can protect the tuning process from failing learners with encapsulation.
The encapsulation separates the tuning from the training of the individual learner.
The encapsulation method is set in the learner.

```{r optimization-030}
learner = lrn("classif.svm",
  cost  = to_tune(1e-5, 1e5),
  gamma = to_tune(1e-5, 1e5),
  kernel = "radial",
  type = "C-classification"
)

learner$encapsulate = c(train = "evaluate", predict = "evaluate")
```

The encapsulation can be set individually for training and predicting.
There are currently two options for encapsulating a learner.
The  `r ref_pkg("evaluate")` package and the `r ref_pkg("callr")` package.
The `r ref_pkg("evaluate")` catches any errors that occur and allows the process to continue.

> The `r ref_pkg("callr")` package comes with more overhead because the encapsulation spawns a separate R process.

Reviewer 3: @be-marc I don't think this has been addressed

Both packages allow setting a timeout which is useful when a learner does not converge.
We set a timeout of 30 seconds.

```{r optimization-031}
learner$timeout = c(train = 30, predict = 30)
```

With encapsulation, exceptions and timeouts do not stop the tuning.
Instead, the error message is recorded and a fallback learner is fitted.

Fallback learners allow scoring a result when no model was fitted during training.
A common approach is to predict a weak baseline e.g. predicting the mean of the data or just the majority class.

> See @sec-fallback for more detailed information.

Reviewer 3: @be-marc this hasn't been addressed - where is sec-fallback?

> See @sec-fallback for more detailed information.

Reviewer 2: @be-marc this has nothing to do with paradox

The `r ref("mlr_learners_classif.featureless", "featureless learner")` predicts the most frequent label.

```{r optimization-032}
learner$fallback = lrn("classif.featureless")
```

Errors and warnings that occurred during tuning are stored in the archive.

```{r optimization-033}
instance = tune(
  method = tnr("random_search", batch_size = 5),
  task = tsk("sonar"),
  learner = learner,
  resampling = rsmp("cv", folds = 3),
  measures = msr("classif.ce"),
  term_evals = 10
)

as.data.table(instance$archive)[, list(cost, gamma, classif.ce, errors, warnings)]
```

### Logarithmic Transformations {#sec-logarithmic-transformations}

> Usually, the `cost` and `gamma` hyperparameters are tuned on the logarithmic scale which means the optimization algorithm searches in $[log(1e-5), log(1e5)]$ but transforms the selected configuration with `exp()` before passing to the learner.

Reviewer 3: @be-marc has this been addressed?

#|   - "`exp(cost)` values seen by the learner."
#| layout-ncol: 2
library(ggplot2)
library(viridisLite)

data = data.frame(cost = cost)
ggplot(data, aes(x = cost)) +
  geom_histogram(
    bins = 15,
    fill = viridis(1, begin = 0.5),
    alpha = 0.8,
    color = "black") +
  theme_minimal()

data = data.frame(cost = exp(cost))
ggplot(data, aes(x = cost)) +
  geom_histogram(
    bins = 15,
    fill = viridis(1, begin = 0.5),
    alpha = 0.8,
    color = "black") +
  theme_minimal()
```

To add the `exp()` transformation to a hyperparameter, we pass `logscale = TRUE` to `to_tune()`.

```{r optimization-053}
learner = lrn("classif.svm",
  cost  = to_tune(1e-5, 1e5, logscale = TRUE),
  gamma = to_tune(1e-5, 1e5, logscale = TRUE),
  kernel = "radial",
  type = "C-classification"
)

instance = tune(
  method = tnr("grid_search", resolution = 5, batch_size = 5),
  task = tsk("sonar"),
  learner = learner,
  resampling = rsmp("cv", folds = 3),
  measures = msr("classif.ce")
)

instance$result
```

The column `x_domain` contains the hyperparameter values after the transformation i.e. `exp(5.76)` and `exp(-5.76)`:

```{r optimization-054}
instance$result$x_domain
```

> ### Creating Search Spaces from Learners {#sec-search-space-token}

Reviewer 3: @be-marc

```{r optimization-046}
learner = lrn("classif.svm")
learner$param_set$set_values(
  kernel = "polynomial", # for example
  degree = to_tune(lower = 1, upper = 3)
)
learner$param_set$search_space()
```

It is possible to omit `lower` here, because it can be inferred from the lower bound of the `degree` parameter itself.
For other parameters, that are already bounded, it is possible to not give any bounds at all, because their ranges are already bounded.
An example is the logical `shrinking` hyperparameter:

```{r optimization-047}
learner$param_set$set_values(shrinking = to_tune())
learner$param_set$search_space()
```

A `r ref("TuneToken")` can also be constructed with a `r ref("Domain")` object, i.e. something constructed with a `p_<type>` call.
This way it is possible to tune continuous parameters with discrete values, or to give transformations or dependencies.
One could, for example, tune the `cost` as above on three given special values, and introduce a dependency of `shrinking` on it.
Notice that a short form for `to_tune(<levels>)` is a short form of `to_tune(p_fct(<levels>))`.

:::{.callout-note}

> When introducing the dependency, we need to use the `degree` value from *before* the implicit transformation, which is the name or `as.character()` of the respective value, here `"val2"`!

Reviewer 2: ```suggestion
When introducing the dependency, we need to use the `cost` value from *before* the implicit transformation. Concretely, we write the dependency as `cost == "val2"` and not `cost == 0.7`.
```

Reviewer 3: It's still confusing but is made clearer with the example

> :::

Reviewer 2: Done

    geom_point()
```

The available types of search space parameters are limited: continuous, integer, discrete, and logical scalars.
There are many machine learning algorithms, however, that take parameters of other types, for example vectors or functions.
These can not be defined in a search space `r ref("ParamSet")`, and they are often given as `r ref("ParamUty")` in the `r ref("Learner")`'s `r ref("ParamSet")`.^[Recall that a parameter of type `r ref("ParamUty")` cannot be used in a search space because it is never bounded.]
When trying to tune over these hyperparameters, it is necessary to perform a Transformation that changes the type of the parameter.

An example is the `class.weights` parameter of the [Support Vector Machine](https://machinelearningmastery.com/cost-sensitive-svm-for-imbalanced-classification/) (SVM), which takes a named vector of class weights with one entry for each target class.
The transformation that would tune `class.weights` for the `r ref("mlr_tasks_spam", text = "spam")` dataset could be:

```{r optimization-041}
search_space = ps(
  class.weights = p_dbl(lower = 0.1, upper = 0.9,
    trafo = function(x) c(spam = x, nonspam = 1 - x))
)

> ```

Reviewer 3: I think this is a bit confusing following from the previous example (also subsequent examples below). Why sometimes `.extra_trafo` and others `trafo`. Why not use vector example again? (these are rhetorical Qs I know the answer) I think the flow could be a bit clearer, maybe something like "We will now look at common advanced tuning methods" then you can talk about vectors, characters, etc. But just framed as problem/solution

Author: @sebffischer

Reviewer 2: Maybe a graphic would be nice, which illustrates how the trafos are first applied and then the .extra_trafo is applied to all the transformed parameters?

Reviewer 3: @be-marc

Reviewer 2: I added an explanation

A common use-case is the necessity to specify a list of values that should all be tried (or sampled from).
It may be the case that a hyperparameter accepts function objects as values and a certain list of functions should be tried.
Or it may be that a choice of special numeric values should be tried.
For this, the `r ref("p_fct")` constructor's `level` argument may be a value that is not a `character` vector, but something else.
If, for example, only the values 0.1, 3, and 10 should be tried for the `cost` parameter, even when doing random search, then the following search space would achieve that:

```{r optimization-042}
search_space = ps(
  cost = p_fct(c(0.1, 3, 10)),
  kernel = p_fct(c("polynomial", "radial"))
)
```

This is equivalent to the following:
```{r optimization-043}
search_space = ps(
  cost   = p_fct(c("0.1", "3", "10"),
    trafo = function(x) list(`0.1` = 0.1, `3` = 3, `10` = 10)[[x]]),
  kernel = p_fct(c("polynomial", "radial"))
)
```

> This may seem silly, but makes sense when considering that factorial tuning parameters are always `character` values:

Reviewer 3: Never say "this may seem silly" about your own code

Author: @sebffischer

> This may seem silly, but makes sense when considering that factorial tuning parameters are always `character` values:

Reviewer 2: ```suggestion
Keep in mind, that factorial tuning parameters are always `character` values:
```

```{r optimization-044}
search_space = ps(
  cost   = p_fct(c(0.1, 3, 10)),
  kernel = p_fct(c("polynomial", "radial"))
)
typeof(search_space$params$cost$levels)
```

:::{.callout-warning}
Be aware that this results in an "unordered" hyperparameter.
Tuning algorithms that make use of ordering information of parameters, like genetic algorithms or model based optimization, will perform worse when this is done.
For these algorithms, it may make more sense to define a `r ref("p_dbl")` or `r ref("p_int")` with a more fitting transformation.
:::

> It is possible to specify dependencies between parameters.

Reviewer 3: This has already been discussed above

Author: @sebffischer

Reviewer 3: @be-marc

Reviewer 2: I removed this repetition

### Logarithmic Transformations {#sec-logarithmic-transformations}

> After the previous section has covered how to make the tuning process more robust to failures, we will now show how to exercise more fine-grained control over the sampling distribution.

Reviewer 3: ```suggestion
We will now look at another advanced topic, which is using transformations to exercise more fine-grained control over the sampling distribution.
```

> In `r ref_pkg("mlr3")`, the standard way to modify the sampling distribution -- besides changing the tuner from e.g. grid search to random search -- is to apply a transformation.

Reviewer 3: ```suggestion
```

> To show how this works, we use the SVM's `cost` and `gamma` parameters that we have tuned in the example above.
> These are usually not tuned on a linear, but logarithmic scale.

Reviewer 3: ```suggestion
To show how this works, we use the SVM's `cost` and `gamma` parameters that we have tuned in the example above, these are usually tuned on a logarithmic (as opposed to linear) scale.
```

### Creating Search Spaces from Learners {#sec-search-space-token}

In @sec-tuning-instance we have seen how one can conveniently define the tuning space of a learner using a `r ref("TuneToken")` that is constructable with the `r ref("to_tune")` function.
In this section, we will provide a more detailed explanation of the underling mechanism and its capabilities.

Hyperparameters can be set to the `r ref("TuneToken")` in the same way that they are set to specific values.

> It can be understood as the hyperparameters being tagged for later tuning.

Reviewer 3: I don't know what this means sorry

It is even possible to define whole `r ref("ParamSet")`s that get tuned over for a single parameter.
This may be especially useful for vector hyperparameters that should be searched along multiple dimensions.
This can be done using an `.extra_trafo`, which is similar to the `logscale` transformation from earlier, but is applied to all parameters.
It can be set to a function with arguments `x` and `param_set` -- the parameter values and the `r ref("ParamSet")` itself -- and must return the transformed parameters.
You will learn more about it in section @sec-paradox.

> In this case special case, the `.extra_trafo` must return a list with a single element, because it corresponds to a single hyperparameter that is being tuned.

Reviewer 3: ```suggestion
In this special case, the `.extra_trafo` must return a list with a single element, as it corresponds to a single hyperparameter that is being tuned.
```

learner$param_set$search_space()
```

> :::{.callout-warning}

Reviewer 3: ```suggestion
::{.callout-tip}
```

It is even possible to add or remove parameters.

In our example we now assume that the parameter `cost` should be set to a higher value when the `kernel` is `"polynomial."`.

> Such a transformation is only possible via a `.extra_trafo`, because it accesses information from multiple parameters at once.

Reviewer 3: Have we defined `.extra_trafo` already? If not this won't make sense
